### ISSUE TITLE
Seems to not always behave correctly with special characters

### ISSUE THREAD

**Issue by rmbasilio** [arlinsandbulte/Pro7-Media-Sweeper#4]
I had a few files that were incorrectly removed. I'm not too sure yet what the problems are, but I suspect it's special characters, like "ñ", or numbers or apostrophes immediately around whitespace.

Some examples of titles that I know for sure were incorrectly removed:

WAY-ALPHA2019-Español-Episodio-01.mp4
02 Jesu, Joy Of Man's Desiring.mp3
Formed, Roman Roads - Spanish.mp4

**Comment by arlinsandbulte** [comment]
We've tested it with special characters, and I am not aware of that type of issue.
Are you on Windows or Mac?
Can you attach the log file from your sweep?
I might also want to see the *.pro file or playlist file that these media files should be referenced in.

**Comment by rmbasilio** [comment]
I'm on Mac. I'll attach the log and the two pro/playlist files of the first two, since I remember for sure which ones those were in.

[Issue4.zip](https://github.com/arlinsandbulte/Pro7-Media-Sweeper/files/7874158/Issue4.zip)

**Comment by arlinsandbulte** [comment]
@rmbasilio
Thanks for sending that to me.  It's been very helpful.
I was able to reproduce your issue.  A new v2.0-RC2 release has been published which should take care of most of your issues.
BUT, there is another issue I found looking at your information that I still need to figure out and fix.  So, I am going to leave this issue open.
I doubt that second part will cause problems for you.
Please try out the v2.0-RC2 release and let me know if you still have issues.

Thanks again!

**Comment by rmbasilio** [comment]
I hate to be the bearer of bad news, but I downloaded RC2 and a lot of the same files still got incorrectly moved over.
If I have some time I'll see if I can run the sourcecode directly on my machine and see if I can find the problem. It's been a long time since I've run python on a machine, so I'll have to setup a IDE and such.

Thanks though for all your work on this!

**Comment by arlinsandbulte** [comment]
:-( 
Hi @rmbasilio good to know you are a coder!  (I'm not really, but google and some friends have been a big help)
I've been trying to work on this all morning (and way too late last night!).
The issue lies with how the path reference strings get encoded (obviously).
@greyshirtguy pointed me here: https://medium.com/swlh/unicode-string-comparisons-and-compositions-many-forms-same-meaning-775487f5d76c
From that, I tried using `unicodedata.normalize()`, but that is not doing what I want or expect.

If you peek in the log file, you will find things like this:
```
Found Media Files: 
--: /Users/arlin/Documents/ProPresenter/Media/Assets/WAY-ALPHA2019-Español-Episodio-01.mp4
Media References in: /Users/arlin/Documents/ProPresenter/Libraries/Default/!Test.pro
--Absolute: file:/Users/arlin/Documents/ProPresenter/Media/Assets/WAY-ALPHA2019-Español-Episodio-01.mp4
--Path: Media/Assets/WAY-ALPHA2019-Espan\314\203ol-Episodio-01.mp4
```

Notice how the found media file text string uses 'ñ' and the Absolute Reference string found in Pro7 uses 'ñ' too.  When I compare those it works.
But, the Path Reference string found in Pro7 uses some sort of escape encoding.... 'ñ' is replaced with 'n\314\203'

I think that is the issue you are finding.

I'm using PyCharm as my IDE.  If you wish, I can share my whole PyCharm project with you so you don't need do hunt down all the dependencies.
Thanks for your help!

**Comment by rmbasilio** [comment]
That would be useful! I'll admit it's been a couple years since I've had to do any real coding (I used to work as a programmer, but now I'm in the AV industry), so I might be a bit rusty, but hopefully I can be helpful soon!

**Comment by arlinsandbulte** [comment]
I think I found how partially fix your issue (again...).  I may need to add unicode.normalize() to absolute paths too.
This 'ñ' is a single character.
This 'ñ' is actually encoded as 2 characters, the second above the first.
They both look the same, but a string comparison '==' between those two examples above will result in FALSE.

It gets confusing in the code when they can look the same, but be different.

**Comment by rmbasilio** [comment]
I could see that being the case. It's weird that wouldn't be consistently one or the other seeing how it's the same file and the same programs writing it, but you never know.

**Comment by arlinsandbulte** [comment]
Right.  That is what I am trying to work through right now.

**Comment by arlinsandbulte** [comment]
If you want to dig in, feel free.
Attached is the PyCharm project folder I have been working with.
I'm on Windows, but I think you can bring this right into PyCharm for Mac too
[Pro7 Media Sweeper.zip](https://github.com/arlinsandbulte/Pro7-Media-Sweeper/files/7875530/Pro7.Media.Sweeper.zip)
.

**Comment by rmbasilio** [comment]
I think I figured out the issue, though I'm still working on the solution.

Some context I should probably disclose: I actually was one of the developers of ProPresenter back when they were first starting on it. I try not to flaunt it on the Facebook group because I don't want to make it look like that I speak for RenewedVision when I don't work there anymore, and it's been a few years, so there's definitely plenty of things they would've changed that I wouldn't know about. But some of the ideas I'm pulling here are coming from some vague memories of the underlying code structure that I think I remember.

So, in `get_refs_in_file()`, you're pulling three possible things, right? `absolute_string`, `relative_path`, and `path`. I think the issue here is that we're treating these as basically all the same thing, but they're actually different. `absolute_string` is a string that can be copy pasted anywhere and would work, whereas `relative_path` and `path` are actually meant to be part of a URL object, so they're encoded differently. It normally doesn't matter with regular characters, but it could cause issues with special characters.

Also, the reason why it wasn't working for me but was for you was probably because it was able to match `absolute_string` for you, but not for me. Because my library being complicated and moved around many computers, some of my absolute paths have the wrong username in the path.

**Comment by arlinsandbulte** [comment]
WOW, small world!  I was going to ask you about the absolute/relative paths... I suspected you might be relying more on the relative paths.
I've made some progress too, figuring out how to better decode the text I am getting from the Pro7 files so it can be compared to the found media file paths.
Here is what I am doing now:
```
absolute_refs = re.findall(absolute_ref_regex, file_obj.__str__())
    for i in range(len(absolute_refs)):
        absolute_refs[i] = unquote(absolute_refs[i])  # Convert ref from url encoding with % codes to plain text
        absolute_refs[i] = eval(f"b'{absolute_refs[i]}'").decode('utf-8')  # Decode escaped chars to utf-8
        absolute_refs[i] = Path(absolute_refs[i])  # Convert string to Path object
        write_file_line(log_file, "--Absolute: " + absolute_refs[i].__str__())
```

**Comment by rmbasilio** [comment]
Ok, I figured out that if you grab the string from the actual URL object, it parses it correctly. I'll see if I can write up the code.

**Comment by arlinsandbulte** [comment]
This is pretty ugly.  But, it works, I think.
[Pro7_Media_Sweeper.py.zip](https://github.com/arlinsandbulte/Pro7-Media-Sweeper/files/7876128/Pro7_Media_Sweeper.py.zip)

**Comment by rmbasilio** [comment]
sorry, I didn't see your new file. But I got it working, though with a little bit bigger changes than I intended. I ended up rewriting the reference grabbing to look at the protobuf objects themselves instead of looking at the text of the raw file. I'll make a pull request so you can take a look at what I changed

**Comment by arlinsandbulte** [comment]
Hi Robert,
I just pushed an update that should totally fix this issue.
I used a different different approach than the code I attached to this issue a couple of comments above.
If you can give it a try and let me know how it goes.  If you want a packaged executable, let me know.
If this goes well, I'll be publishing a new RC release soon.
THANKS FOR YOUR HELP!
https://github.com/arlinsandbulte/Pro7-Media-Sweeper/commit/e82d13f6505a23fe1b6bc2c4900cd2fbf513d388#diff-e4dcc6e4986249fa211e02f3629baefddf9c459ac054b54f7aa5022c37eacc68

**Comment by arlinsandbulte** [comment]
I just published v2.0-RC3, which should take care of this issue.
It still uses the original method to get reference strings from the Pro7 file string instead of the Pro7 file object.  I will work on that later.
